### PR TITLE
Blog app: add authorization rules

### PR DESCRIPTION
**Pull request by SanaTameem**
# [Project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_add_authorization.md#project-requirements)

- Install CanCanCan in your project.
- Add a role column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
  - For that you need to implement the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
  - For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

File: app/views/posts/show.html.erb [hunk 1]
```diff
@@ -1,15 +1,25 @@
...
+<h1 class = "page-title">Show a specific post</h1>
 <div class = "specific-post-container">
     <div class = "specific-user-posts">
                 <div class = "user-post">
                     <div class ="user-post-counters">Comments: <%=@post.comments_counter%>, Likes: <%=@post.likes_counter%></div>
                     <h3 class ="user-post-title"><%=@post.title%> by <%=@user.name%></h3>
                     <p class ="user-post-text"><%=@post.text%></p>
+                    <%if @post.author == current_user || can?(:destroy, @post)%>
+                        <%= button_to 'Delete', user_post_path(user_id: @post.author_id, id: @post.id), method: :delete, data: {confirm: 'Are you sure you want to delete this post ?'} %>
+                    <% end %>
```
Comment (ridwanediallo) [inline app/views/posts/show.html.erb]: - [ ] Well done implementing the authorization feature 👏. However, please let's ensure that the `Delete` button is only visible if the signed-in user is an `admin` or the `owner` of the post. :ok_hand: 
At this moment, the `Delete` button is visible to unauthorized users, which violates one of the project requirements. Kindly, let's fix that.

### Requirement :point_down: 
Screenshot from 2023-10-20 23-03-14

File: app/views/posts/show.html.erb [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
                 </div>
                 <div class = "post-comments-container">
                     <%@post.comments.each do |comment|%>
-                        <p><%=@user.name%>: <%=comment.text%></p>
+                        <p><%=comment.user.name%>: <%=comment.text%></p>
+                        <% if comment.user == current_user || can?(:destroy, comment)%>
+                            <%= button_to 'Delete Comment', user_post_comment_path(user_id: comment.user_id, post_id: comment.post_id, id: comment.id), method: :delete, data: { confirm: 'Are you sure you want to delete this comment?' }, class: "delete-comment-btn"%>
```
Comment (ridwanediallo) [inline app/views/posts/show.html.erb:16]: - [ ] Well done implementing the authorization feature 👏. However, please let's ensure that the `Delete` button is only visible if the signed-in user is an `admin` or the `owner` of the comment. :ok_hand: 
At this moment, the `Delete` button is visible to unauthorized users, which violates one of the project requirements. Kindly, let's fix that.


### Requirement :point_down: 
Screenshot from 2023-10-20 22-43-05

**Review by ridwanediallo — CHANGES_REQUESTED** [review]
# Changes Required ♻️
## Hi team, 
AlmostThereReachGIF

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used GitHub flow correctly. ✔
- The authorization rules have successfully been set up. ✔


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by rloterh — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi Sana,  

Your project is complete! 🏁 There is nothing else to say other than... it's time to merge it 🍾 Congratulations!  🎉💯🌟 

## Highlights 🎯 
🎲 Good job on implementing recommended changes!
🎲 ReadMe looks great!

## Optional suggestions 🧰 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

**You can also consider:**
Let's make the ReadMe stand out even better by including instructions under the _Getting Started_ section, on how to run the app, for instance, using the command `rails s`.

Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by SanaTameem** [comment]
@ridwanediallo 
@rloterh 
Thank you for reviewing our project and for taking the time to provide your valuable feedbacks. I sincerely appreciate it.

**Pull request merged**